### PR TITLE
Sustainable days

### PR DESCRIPTION
**Pull request by EliasEriksson**
* speed up of `Cadence.allocated(...)`
* implemented `Cadence.sustained(...)`
  * [description](https://github.com/issuefab/app/issues/232)

### CODE REVIEW

File: Cadence/index.ts [hunk 1]
```diff
@@ -7,32 +7,121 @@ export interface Cadence extends Amount {
...
+				const t = Math.abs(
+					(new Date(self.created).getTime() - new Date(cadence.created).getTime()) / 1_000 / 3_600 / 24
+				)
+				const c = costPerDayPerCadence[index]
+				return result + c * t
+			}, 0)
+		const denominator = costPerDayPerCadence.reduce((result, cost) => result + cost, 0)
+		const result = numerator / denominator
+		return Math.trunc(result)
+	}
```
Comment (EliasEriksson) [inline Cadence/index.ts:74]: rename the one letter variables

File: CostCenter/index.ts [hunk 2]
```diff
@@ -54,6 +54,7 @@ export namespace CostCenter {
 					root => (result = (result => (!result ? result : { ...result, root }))(remove(root.costCenters, id)))
 			  ) && result
 	}
+	// validation must chest that all children occurs after created date
```
Comment (EliasEriksson) [inline CostCenter/index.ts:57]: This will be fixed by this issue https://github.com/issuefab/app/issues/250

File: Cadence/index.ts [hunk 3]
```diff
@@ -7,32 +7,121 @@ export interface Cadence extends Amount {
...
+	export const intervals = ["single", "day", "week", "month", "year"] as const
 	export type Interval = typeof intervals[number]
 	export const type = Amount.type.extend<Cadence>({
 		interval: isly.union(isly.string("year"), isly.string("month"), isly.string("week"), isly.string("single")),
 		created: isly.fromIs("Date", isoly.Date.is),
 	})
 	export const is = type.is
 	export const flaw = type.flaw
-	export function allocated(cadence: Cadence, date: isoly.Date): number {
+	function dayDiff(first: isoly.Date | isoly.DateTime, other: isoly.Date | isoly.DateTime): number {
```
Comment (EliasEriksson) [inline Cadence/index.ts:18]: rename this function

**Review by EliasEriksson — COMMENTED** [review]
self check

File: Cadence/index.ts [hunk 4]
```diff
@@ -7,32 +7,121 @@ export interface Cadence extends Amount {
...
+		cap = Math.max(allocated(self, date), cap ?? 0) - singles.reduce((result, cadence) => result + cadence.value, 0)
+
+		const max = dayDiff(date, self.created)
+		const approximation = Math.max(0, Math.min(max, approximate(self, children, date, { cap })))
+		const approximationDate = isoly.Date.next(self.created, approximation)
+		const childCost = children.reduce((result, cadence) => result + allocated(cadence, approximationDate), 0)
+		const approximateCap =
+			Math.min(allocated(self, approximationDate), cap) - singles.reduce((result, cadence) => result + cadence.value, 0)
 
+		//
```
Comment (TheLillyleaf) [inline Cadence/index.ts:100]: remove lines

**Review by TheLillyleaf — CHANGES_REQUESTED** [review]
minor things.

File: Cadence/index.ts [hunk 5]
```diff
@@ -7,32 +7,121 @@ export interface Cadence extends Amount {
 	created: isoly.Date
 }
 export namespace Cadence {
-	export const intervals = ["single", "week", "month", "year"] as const
+	export const intervals = ["single", "day", "week", "month", "year"] as const
 	export type Interval = typeof intervals[number]
 	export const type = Amount.type.extend<Cadence>({
 		interval: isly.union(isly.string("year"), isly.string("month"), isly.string("week"), isly.string("single")),
```
[the remark this answers is not in the document]
Comment (EliasEriksson) [reply Cadence/index.ts:13]: required for the sustain function. maybe ill should source the intervals from elsewhere.

[the remark this answers is not in the document]
Comment (EliasEriksson) [reply Cadence/index.ts:13]: never mind. i only needed it for testing

**Review by EliasEriksson — COMMENTED** [review]
self check

**Review by TheLillyleaf — APPROVED** [review]

**Pull request merged**